### PR TITLE
Fix SyntaxError in TypeScript

### PR DESCRIPTION
**Pull request by junish**
This pull request is related to #55.

After pull request #53, gettext-extract doesn't work in TypeScript with the following `SyntaxError`:
```
> npx gettext-extract --output locales/messages-template.pot --attribute v-translate --parseScript false src/*.vue src/components/*.vue src/views/*/*.vue

[easygettext] extracting: 'src/App.vue
[easygettext] could not read: 'src/App.vue
Trace: { SyntaxError: Unexpected character '@' (24:0)
    at Parser.pp$4.raise (/Users/junichi/Projects/tenri-ui/node_modules/acorn/dist/acorn.js:2757:13)
    at Parser.pp$8.getTokenFromCode (/Users/junichi/Projects/tenri-ui/node_modules/acorn/dist/acorn.js:4906:8)
    at Parser.<anonymous> (/Users/junichi/Projects/tenri-ui/node_modules/acorn-class-fields/inject.js:41:21)
    at Parser.getTokenFromCode (/Users/junichi/Projects/tenri-ui/node_modules/acorn-private-methods/inject.js:30:21)
    at Parser.pp$8.readToken (/Users/junichi/Projects/tenri-ui/node_modules/acorn/dist/acorn.js:4628:15)
    at Parser.pp$8.nextToken (/Users/junichi/Projects/tenri-ui/node_modules/acorn/dist/acorn.js:4619:15)
    at Parser.pp$8.next (/Users/junichi/Projects/tenri-ui/node_modules/acorn/dist/acorn.js:4576:8)
    at Parser.pp.eat (/Users/junichi/Projects/tenri-ui/node_modules/acorn/dist/acorn.js:577:10)
    at Parser.pp.semicolon (/Users/junichi/Projects/tenri-ui/node_modules/acorn/dist/acorn.js:624:13)
    at Parser.pp$1.parseImport (/Users/junichi/Projects/tenri-ui/node_modules/acorn/dist/acorn.js:1482:8)
    at Parser.pp$1.parseStatement (/Users/junichi/Projects/tenri-ui/node_modules/acorn/dist/acorn.js:801:47)
    at Parser.parseStatement (/Users/junichi/Projects/tenri-ui/node_modules/acorn-dynamic-import/lib/inject.js:47:31)
    at Parser.parseStatement (/Users/junichi/Projects/tenri-ui/node_modules/acorn-import-meta/inject.js:39:50)
    at Parser.pp$1.parseTopLevel (/Users/junichi/Projects/tenri-ui/node_modules/acorn/dist/acorn.js:706:23)
    at Parser.parse (/Users/junichi/Projects/tenri-ui/node_modules/acorn/dist/acorn.js:551:15)
    at Object.parse (/Users/junichi/Projects/tenri-ui/node_modules/acorn/dist/acorn.js:5288:37)
  pos: 916,
  loc: Position { line: 24, column: 0 },
  raisedAt: 916 }
    at /Users/junichi/Projects/tenri-ui/node_modules/easygettext/src/extract-cli.js:76:13
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/Users/junichi/Projects/tenri-ui/node_modules/easygettext/src/extract-cli.js:57:7)
    at Module._compile (internal/modules/cjs/loader.js:689:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
    at Module.load (internal/modules/cjs/loader.js:599:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
    at Function.Module._load (internal/modules/cjs/loader.js:530:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:742:12)
    at findNodeScript.then.existing (/usr/local/lib/node_modules/npm/node_modules/libnpx/index.js:268:14)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! tenri-ui@0.1.0 i18n:extract-template: `npx gettext-extract --output locales/messages-template.pot --attribute v-translate --parseScript false src/*.vue src/components/*.vue src/views/*/*.vue`
npm ERR! Exit status 1
```

To avoid the error, I added the detecting that the script is TypeScript and avoid the parsing.

Could you review this pull request?

### CODE REVIEW

File: src/extract-cli.js [hunk 1]
```diff
@@ -56,21 +56,26 @@ const extractor = new extract.Extractor({
...
+      data = script.content;
+      if (script.lang) {
+        ext = script.lang;
+      }
     }
 
-    extractor.parseJavascript(file, data);
+    if (ext !== 'ts') {
+      extractor.parseJavascript(file, data);
+    }
```
Comment (vperron) [inline src/extract-cli.js]: that seems incorrect to me (you should add tests for the cli !) but at the very least ``extract.preprocessTemplate`` is not called anymore. Is that intentional ?

Also, I'd rather (but it's less important)
- use a construct that enables "script file extensions" instead of a series of ``ext !== 'js'``: ['js', 'ts'].includes(ext) is better.
- try/catch exceptions when trying to parse the Javascript, instead of relying on the extension. It's way more defensive and will allow you to parse exotic extensions ^^ you would not need the second "if"

File: src/extract.js [hunk 2]
```diff
@@ -50,11 +50,23 @@ exports.TranslationReference = class TranslationReference {
...
 
 function preprocessScriptTags(data, type) {
+  if (type !== 'vue') {
+    return {};
+  }
+
   const vueFile = vueCompiler.parse({ source: data, needMap: false });
+  if (!vueFile.script) {
+    return {};
+  }
```
Comment (vperron) [inline src/extract.js]: no :) if you do this you will hide the fact that it failed. 

Raise (or return null but it's not as good), check it and continue the loop in the calling code.

File: src/extract.js [hunk 3]
```diff
@@ -50,11 +50,23 @@ exports.TranslationReference = class TranslationReference {
 };
 
 function preprocessScriptTags(data, type) {
+  if (type !== 'vue') {
+    return {};
+  }
```
Comment (vperron) [inline src/extract.js]: this should be checked before, not after entering the function. Maybe call the function preprocessVueFile and only go there if you actually have a Vue file.

File: src/extract.js [hunk 4]
```diff
@@ -50,11 +50,23 @@ exports.TranslationReference = class TranslationReference {
...
-    ? ''
-    : vueFile.script.content.trim();
+  const script = {};
+  if (vueFile.script.content) {
+    script.content = vueFile.script.content.trim();
+  }
+  if (vueFile.script.lang) {
+    script.lang = vueFile.script.lang.trim();
+  }
+  return script;
```
Comment (vperron) [inline src/extract.js]: I think that if there is a ``vueFile.script`` you will have the lang and the content (or it's weird !)

So just return like
```javascript
return {
  script: vueFile.script.content.trim(),
  lang: vueFile.script.lang,
}
```
why would you trim() the lang ? :)

File: src/test-fixtures.js [hunk 5]
```diff
@@ -105,7 +105,7 @@ exports.HTML_LINEBREAK_FILTER = `
 
 <a href="#"
   ng-click="vm.doSomething()"
-  class="button">{{ 'Multi-line 4' | translate 
+  class="button">{{ 'Multi-line 4' | translate
```
Comment (vperron) [inline src/test-fixtures.js]: all those 4 above patches should go in a separate commit that fixes spaces

**Comment by junish** [comment]
@vperron 
Thank you for your quick review and sorry for my broken pull request 😓 
I'm fixing this pull request with your accurate comment.

File: src/extract.js (hunk 4 above)
Comment (junish) [reply src/extract.js]: > So just return like

Looks better! I changed the code at 262d4e4.

> why would you trim() the lang ? :)

I added it just in case.
It's no need so I removed the trim() for the lang at also 262d4e4.

File: src/extract.js (hunk 2 above)
Comment (junish) [reply src/extract.js]: I addressed this comment and added a test at 57af6f0

File: src/extract.js (hunk 3 above)
Comment (junish) [reply src/extract.js]: I moved out the check for `type` from `preprocessScriptTags` at 9895e63.

Then the function name was changed to `preprocessVueFile` at 5884897.

File: src/extract-cli.js (hunk 1 above)
Comment (junish) [reply src/extract-cli.js]: > extract.preprocessTemplate is not called anymore. Is that intentional ?

It's removed accidentally... I reverted it at c36956d. Sorry! 
Tests for cli should be added.

Comment (junish) [reply src/extract-cli.js]: > ['js', 'ts'].includes(ext) is better.

Thank you! I'll use the code after this ^^

File: src/extract-cli.js [hunk 6]
```diff
@@ -66,11 +66,18 @@ files.forEach(function(filename) {
...
+    if (ext === 'vue') {
+      const script = extract.preprocessVueFile(data);
+      if (script) {
+        data = script.content;
+        lang = script.lang;
+      }
     }
 
-    extractor.parseJavascript(file, data);
+    if (lang === 'js') {
```
Comment (vperron) [inline src/extract-cli.js]: warning: lang will be ``undefined `` in most cases here when you come back from a processed Vue file. We should set it as default to 'js' inside ``preprocessVueFile`` I think.

File: src/extract.js [hunk 7]
```diff
@@ -49,12 +49,15 @@ exports.TranslationReference = class TranslationReference {
...
-function preprocessScriptTags(data, type) {
+function preprocessVueFile(data) {
   const vueFile = vueCompiler.parse({ source: data, needMap: false });
-
-  return (type !== 'vue' || ! vueFile.script)
-    ? ''
-    : vueFile.script.content.trim();
+  if (!vueFile.script) {
+    return null;
+  }
```
Comment (vperron) [inline src/extract.js]: returning null is still worse than using ``raise`` + try/catch, but why not.

File: src/extract.js [hunk 8]
```diff
@@ -49,12 +49,15 @@ exports.TranslationReference = class TranslationReference {
...
+  return {
+    content: vueFile.script.content.trim(),
+    lang: vueFile.script.lang,
```
Comment (vperron) [inline src/extract.js]: I suggest (cf higher in the comments)

```suggestion
    lang: vueFile.script.lang || 'js',
```

to avoid the undefined by default

**Comment by vperron** [comment]
Hello :) If you're not planning on adding CLI tests (a shame, looks like you're in the mood !) I would agree to merge this PR. Can you please squash your commit properly before I do a final read ?

**Comment by junish** [comment]
I will create tests for cli in another pull request. Anyway I squashed this pull request.

**Comment by vperron** [comment]
Thanks for your contribution :)

**Comment by Shin--** [comment]
@junish This fixes crashes when parsing TypesScript but only due to not parsing JS at all, right?
@vperron Are there any plans to add actual TypeScript support? I could look into it, but I would probably have to add an additional dependency since acorn only parses JavaScript.

**Comment by vperron** [comment]
I'd be extremely delighted if you can look into it :) I'm terribly busy at the moment; but I'm working on Typescript as well on my project. I'll gladly read any PR that comes this way, and make you a permanent contributor as soon as it's merged :)
